ide/coqide/preferences.ml

```diff
@@ -321,21 +321,20 @@ let attach_modifiers (pref : string preference) ?(filter = Fun.const true) prefi
   in
   pref#connect#changed ~callback:cb
 
+let select_arch m m_mac =
+  (* Gtk's <Primary> maps to <Meta> (i.e. "Command") on Darwin (i.e. Mac),
+    <Control> on other architectures; but sometimes, we want more distinction *)
+  if Coq_config.arch = "Darwin" then m_mac else m
+
 let modifier_for_navigation =
   new preference ~name:["modifier_for_navigation"]
-    (* Note: on Darwin, this will give "<Control><Meta>", i.e. Ctrl and Command; on other
-    architectures, "<Primary>" binds to "<Control>" so it will give "<Control>" alone *)
-    ~init:"<Control><Primary>" ~repr:Repr.(string)
+    ~init:(select_arch "<Alt>" "<Control><Primary>") ~repr:Repr.(string)
```

Comment:
Same remark about the potential revival of a debate about the default modifier. I'd recommend not touching it.

Reply:
Having the classic Ctrl-Right, Ctrl-Home and Ctrl-End not working by default gives the impression that CoqIDE is not a serious IDE, to my mind. Better is to align to what VsCoq offers. The debate for the modifier in VsCoq was probably better.

Reply:
The problem is that the availability of these bindings really depend on your desktop environment, and you'd be surprised by the amount of variability in our end users. I don't know about the defaults for VsCoq, do you have a pointer?

> the impression that CoqIDE is not a serious IDE

CoqIDE is a terrible text editor, and I think that any reasonable person that has ever used or maintained it would agree with that. :shrug:

Reply:
Ctrl-Right, Ctrl-Home and Ctrl-End are mapped by the editor (GtkSourceView), not the desktop environment. Therefore they are _always_ unavailable.

The modifier for navigation in VsCoq is Alt, at least on Windows and Linux. I tested it at the university. On Mac, Alt-Right is the equivalent of Ctrl-Right of other systems so maybe another choice was made. I don’t know if this setting can be changed in VsCoq.

Reply:
On Mac, VsCoq 1 and the manual mode of VsCoq 2 use Ctrl+Alt. VsCoq shortcuts are configurable. It has Ctrl-Alt-End and F8 to the end, Ctrl-Alt-Right and F5 to the current point, Ctrl-Alt-Up and Ctrl-Alt-Down to step, but Ctrl-Alt-Home does not seem to be bound. Regarding coq-lsp, there is (apparently) no such bindings (only an asynchronous evaluation mode).

Reply:
I changed my mind.  We really should support the widely used Ctrl-Right, Ctrl-Home and Ctrl-End functions.  (Ctrl-Left works, BTW.)  Given that key bindings associated with menus can be rebound (see [Key Bindings](https://coq.inria.fr/doc/V8.19.0/refman/practical-tools/coqide.html#key-bindings in the doc), I think switching the Navigation menu to use Alt is reasonable provided that:
- the doc fully states what bindings have been changed and briefly explains why we're making the change (users will accept the change much more readily if they understand why)
- describes how to reverse the change by editing the key bindings, including the exact lines to add to the key bindings file to continue using the current Ctrl bindings
- direct CoqIDE users attention to this in the changelog and release notes.

The impact will be small--a quick, one-time fix for those who want to keep the old bindings.  And easy enough for everyone else to adapt to the new bindings, which are shown in the menus.

BTW, does vsCoq support Ctrl-Right, Ctrl-Home and Ctrl-End?

Reply:
Changing the modifier is actually even simpler than you think: you can go to Edit > Preferences > Shortcuts, and change the modifier for Navigation menu. It is also very easy to notice that the shortcuts have changed for the Navigation menu. Also note that is already stated in the doc that Ctrl-Home and Ctrl-End move the cursor “if you remove the default bindings” (changing the default modifier to Alt allows to remove this condition).

Therefore we should write the Customizations > Shortcuts section of the doc. I’ve planned to work on it in another pull request…

Where should I put the note about the change of modifier? In the section “Running Coq scripts,” just below the shortcuts?

Reply:
I’ve put the note in the doc.

Reply:
Hmm, when I run your branch on my system (WSL on Windows, a Linux VM), the Nav menu items still use Ctrl rather than Alt.

FYI I will be travelling until April 10 and won't be able to investigate this until then.

Reply:
This is because it is loaded from the preferences file, coqiderc, located at ~/.config/coq/coqiderc on my system. If you want to restore the default modifier, you have to remove the line with `modifier_for_navigation`.

Reply:
I think line 331 should use "<Alt>" regardless of the platform, then the changelog should include a note explaining why (to free up hidden Control shortcuts) and telling how to change it to "<Control>" from Preferences/Shortcuts.  Consistency on the key bindings is simpler.

As you point out, this won't change anything if the config files already exist.  I think it's worth nudging users to switch to the new defaults by changing the modifier.  Perhaps you could create a new preference, e.g. "changed_nav_default" after resetting the modifier so it's not every time you start CoqIDE.

We should the doc so it says the correct things about the Ctrl- combinations it mentions.  Also,
the documentation should say what the Preferences/Shortcuts menu does.

WDYT?

Reply:
> I think line 331 should use "<Alt>" regardless of the platform, then the changelog should include a note explaining why (to free up hidden Control shortcuts) and telling how to change it to "<Control>" from Preferences/Shortcuts.

Alt is not a good choice for Mac, as Alt-Right is the equivalent of Ctrl-Right on Unix and Windows. As @herbelin said, the shortcut on VsCoq is Ctrl-Alt. Should we switch to Ctrl-Alt?

>  I think it's worth nudging users to switch to the new defaults by changing the modifier automatically. Consistency on the key bindings is simpler. They can change back to "<Control>" if they like. Perhaps you could create a new preference, e.g. "changed_nav_default" after resetting the modifier so it's not reset every time you start CoqIDE.

What matters more to me is the case for new users. People who already use CoqIDE probably know the preference and should already have switched to the Alt modifier.

> Then we should update the doc so it says the correct things about the Ctrl- combinations it mentions. Also, the documentation should say what the Preferences/Shortcuts menu does.

I think I’ve done it already, haven’t I?

> It's not sufficient to remove `modifier_for_navigation` from the config file

It was sufficient for me (CoqIDE must be closed when you edit the config file). But it was just for testing; of course, it’s easier to do the change from the preferences dialog.

Reply:
> - I think vsCoq compatibility is not very relevant, so I would just stick with Alt rather than use Ctrl+Alt.

I guess you mean Ctrl+Cmd.

Ctrl-Z is a much more common key binding for undo than Ctrl-U.

> WDYT?

To me, things are OK…

Reply:
I see Ctrl-Z in notepad and MS Word on Windows.  In Linux, Ctrl-Z suspends the process.  Is it used on Mac?

Reply:
>  In Linux, Ctrl-Z suspends the process.

This is the case only in the terminal. In GUI applications, Ctrl-Z is used for undo. See editors such as IntelliJ IDEA, VS Code, Mousepad; but also image editing software such as GIMP and Inkscape. On Mac, the equivalent is Cmd-Z.

Furthermore, Ctrl-Z and Ctrl-Y are already bound to undo and redo by the GtkSourceView editor in CoqIDE actually.